### PR TITLE
Add unit tests for `OAHashMap`

### PR DESCRIPTION
**Pull request by Nazarwadim**

### CODE REVIEW

File: tests/core/templates/test_oa_hash_map.h [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+			Ref<Texture2D> ref_texture_2d;
+
+			map1.set(string, ref_texture_2d);
+			Ref<Texture2D> map_vec = *map1.lookup_ptr(string);
+			CHECK(map_vec == ref_texture_2d);
+		}
+		OAHashMap<String, Ref<Texture2D>> map1copy(map1);
+		CHECK(map1copy.has(String("qwerty") + '0'));
+		map1copy = map1;
+		CHECK(map1copy.has(String("qwerty") + 2 + '0'));
```
Comment (AThousandShips) [inline tests/core/templates/test_oa_hash_map.h:167]: What are you trying to achieve here? You're adding a character value, did you mean
```suggestion
		CHECK(map1copy.has(String("qwerty") + '2' + '0'));
```

Comment (Nazarwadim) [reply tests/core/templates/test_oa_hash_map.h:167]: This should be  `CHECK(map1copy.has(String("qwerty") + '2'));`

Comment (Nazarwadim) [reply tests/core/templates/test_oa_hash_map.h:167]: Or better `CHECK(map1copy.has(String("qwerty") + (2 + '0')));`

Comment (AThousandShips) [reply tests/core/templates/test_oa_hash_map.h:167]: Why though? It should just be `CHECK(map1copy.has(String("qwerty2"))`

File: tests/core/templates/test_oa_hash_map.h [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+}
+
+TEST_CASE("[OAHashMap] Non-trivial types") {
+	uint64_t pre_mem = Memory::get_mem_usage();
+	{
+		OAHashMap<String, Ref<Texture2D>> map1;
+		const uint32_t count = 10;
+		for (uint32_t i = 0; i < count; i++) {
+			String string = "qwerty";
+			string += i + '0';
```
Comment (AThousandShips) [inline tests/core/templates/test_oa_hash_map.h:157]:
```suggestion
			string += itos(i) + '0';
```
Or you'll add null and other characters, there's no reason to test that

Comment (AThousandShips) [inline tests/core/templates/test_oa_hash_map.h:157]: I'd just use `itos` this is very cryptic

File: tests/core/templates/test_oa_hash_map.h [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+	CHECK(value == 84);
+	CHECK(data == 84);
+}
+
+TEST_CASE("[OAHashMap] Overwrite element") {
+	OAHashMap<int, int> map;
+	map.set(42, 84);
+	map.set(42, 1234);
+	int result = *map.lookup_ptr(42);
+	CHECK(result);
```
Comment (Calinou) [inline tests/core/templates/test_oa_hash_map.h:68]: This should check the new value as well:

```suggestion
	CHECK(result == 1234);
```

File: tests/core/templates/test_oa_hash_map.h [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+	expected.push_back(Pair<int, int>(42, 84));
+	expected.push_back(Pair<int, int>(123, 111111));
+	expected.push_back(Pair<int, int>(0, 12934));
+	expected.push_back(Pair<int, int>(123485, 1238888));
+
+	int idx = 0;
+	for (OAHashMap<int, int>::Iterator it = map.iter(); it.valid; it = map.next_iter(it)) {
+		int64_t result = expected.find(Pair<int, int>(*it.key, *it.value));
+		CHECK(result >= 0);
+		++idx;
```
Comment (Calinou) [inline tests/core/templates/test_oa_hash_map.h:107]: Prefer post-incrementation if you don't rely on pre-incrementation behavior:

```suggestion
		idx++;
```

**Review by Calinou — COMMENTED** [review]
Looks good to me, please squash the commits together so we can merge this. See [PR workflow](https://docs.godotengine.org/en/latest/community/contributing/pr_workflow.html#mastering-the-pr-workflow-the-rebase) for instructions :slightly_smiling_face:

**Review by Calinou — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

File: tests/core/templates/test_oa_hash_map.h [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+	map.insert(123485, 1238888);
+	map.set(123, 111111);
+
+	Vector<Pair<int, int>> expected;
+	expected.push_back(Pair<int, int>(42, 84));
+	expected.push_back(Pair<int, int>(123, 111111));
+	expected.push_back(Pair<int, int>(0, 12934));
+	expected.push_back(Pair<int, int>(123485, 1238888));
+
+	int idx = 0;
```
[the remark this answers is not in the document]
Comment (akien-mga) [reply tests/core/templates/test_oa_hash_map.h:103]: Weird that it wasn't caught on CI. Which compiler and compile option do you use?

[the remark this answers is not in the document]
Comment (vnen) [reply tests/core/templates/test_oa_hash_map.h:103]: I'm using Clang with `dev_mode=yes`.

**Pull request merged**